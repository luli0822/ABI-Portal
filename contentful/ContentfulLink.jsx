import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';
import siteDataContext from './siteDataContext';
import { linkPropType } from './prop-types';
import { fallbackLang } from './config';

/**
 * Build a link component that will render either a pageLink or externalLink
 * @param {*} props
 */
export default function ContentfulLink(props) {
	const { linkEntry, children, targetLang, ...aProps } = props;

	// check the type for this link. it's internal if the type is pageLink. otherwise it's externalLink.
	const isInternal = linkEntry.sys.contentType.sys.id === 'pageLink';

	const { pathMap, lang, fieldOrFallback } = useContext(siteDataContext);

	const finalLang = targetLang || lang;

	const ariaLabel = fieldOrFallback(linkEntry.fields.accessibleLabel, finalLang);

	if (isInternal) {
		const destinationPageId = linkEntry.fields.page[fallbackLang].sys.id;

		const hash = linkEntry.fields.hash ? `#${linkEntry.fields.hash[finalLang]}` : '';

		// TODO: find strategy for handling links to pages that don't exist in the path map.
		const [asPath, { page }] = Object.entries(pathMap).find(
			([path, options]) =>
				options.query.pageId === destinationPageId && options.query.lang === finalLang
		);

		const href = `${page}?pageId=${destinationPageId}&lang=${finalLang}${hash}`;

		return (
			<Link href={href} as={asPath}>
				{/* eslint-disable-next-line react/jsx-props-no-spreading */}
				<a aria-label={ariaLabel} {...aProps}>
					{children}
				</a>
			</Link>
		);
	}

	const href = fieldOrFallback(linkEntry.fields.url, finalLang);

	return (
		// eslint-disable-next-line react/jsx-props-no-spreading
		<a href={href} aria-label={ariaLabel} {...aProps}>
			{children}
		</a>
	);
}

ContentfulLink.propTypes = {
	linkEntry: linkPropType.isRequired,
	children: PropTypes.node.isRequired,
	targetLang: PropTypes.string,
};

ContentfulLink.defaultProps = {
	targetLang: undefined,
};
