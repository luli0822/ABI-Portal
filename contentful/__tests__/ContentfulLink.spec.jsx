import React from 'react';
import { shallow } from 'enzyme';
import ContentfulLink from '../ContentfulLink';
import {
	mockLang,
	mockPathMap,
	mockLinkedEntries,
	mockFieldOrFallback,
} from '../__mocks__/SiteDataContext';

// Mock React useContext
jest.mock('react', () => {
	const ActualReact = require.requireActual('react');
	return {
		...ActualReact,
		useContext: () => ({
			pathMap: mockPathMap,
			lang: mockLang,
			fieldOrFallback: mockFieldOrFallback,
		}),
	};
});

const mockPageLinkEntry = mockLinkedEntries['about-us-page-link'];

const mockExternalLinkEntry = mockLinkedEntries['some-external-link'];

describe('<ContentfulLink />', () => {
	it('Renders a pageLink correctly - snapshot', () => {
		const component = shallow(
			<ContentfulLink linkEntry={mockPageLinkEntry}>Page Link Text</ContentfulLink>
		);
		expect(component.debug()).toMatchSnapshot();
	});

	it('Check a next.js Link is rendered for page link', () => {
		const component = shallow(
			<ContentfulLink linkEntry={mockPageLinkEntry}>Page Link Text</ContentfulLink>
		);

		expect(component.find('Link[href]')).toHaveLength(1);
		expect(component.find('a')).toHaveLength(1);
		expect(component.find('a[href]')).toHaveLength(0);
	});

	it('Renders an external link correctly - snapshot', () => {
		const component = shallow(
			<ContentfulLink linkEntry={mockExternalLinkEntry}>Externsl link text</ContentfulLink>
		);
		expect(component.debug()).toMatchSnapshot();
	});

	it('Check an <a> tag is rendered on enternal link', () => {
		const component = shallow(
			<ContentfulLink linkEntry={mockExternalLinkEntry}>External link text</ContentfulLink>
		);
		// For external link, there is an <a> tag and no next Link
		expect(component.find('a[href]')).toHaveLength(1);
		expect(component.find('Link')).toHaveLength(0);
	});
});
