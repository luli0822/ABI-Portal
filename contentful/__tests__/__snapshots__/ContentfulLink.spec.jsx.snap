// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentfulLink /> Renders a pageLink correctly - snapshot 1`] = `
"<Link href=\\"/?pageId=about-us-page&lang=en-CA#s\\" as=\\"/about-pg-en\\">
  <a aria-label={[undefined]}>
    Page Link Text
  </a>
</Link>"
`;

exports[`<ContentfulLink /> Renders an external link correctly - snapshot 1`] = `
"<a href=\\"https://test.com\\" aria-label=\\"some accessibility text\\">
  Externsl link text
</a>"
`;
